text/0000-synthetic-shadow-migrate.md

```diff
@@ -0,0 +1,140 @@
+
+The most serious incompatibility between native and synthetic shadow DOM is the lack of global styling for native shadow components. In synthetic shadow, any global stylesheets still affect elements inside the shadow root. As such, many component authors have taken dependencies on global stylesheets, notably on [SLDS](https://www.lightningdesignsystem.com/) in Lightning Experience. In fact, the Salesforce documentation explicitly [encourages this](https://developer.salesforce.com/docs/platform/lwc/guide/create-components-css-slds.html) in "getting started" guides, and community-authored content also frequently promotes `slds-*` classes as well.
+
+This RFC proposes a new mode to help ease this migration, and to encourage more component authors to move from synthetic to native shadow, by emulating synthetic shadow's styling behavior on top of native shadow DOM.
+
+## Detailed design
+
+`shadowSupportMode` has two options – `'reset'` (synthetic) or `'native'` (native). Currently on Lightning Experience,
+`'reset'` is the default, but developers may opt into `'native'`, which 1) uses native shadow DOM, and 2) cascades native
+shadow DOM down into any descendants of the component.
+
+This RFC proposes a third mode: `'migrate'`:
+
+```js
+static shadowSupportMode = 'migrate'
+```
+
+"Migrate" mode (ala [jquery-migrate](https://github.com/jquery/jquery-migrate)) has the following behaviors:
+
+- Native shadow DOM is used (like `'native'`).
+- The mode cascades to all descendants (like `'native'`).
+- Unlike `'native'`, however, "migrate" mode copies global `<style>`s and `<link rel="stylesheet">`s from the `document`'s `<head>` into the shadow root of the component. It does so once, when the component is connected to the DOM.
+
+By copying the styles from the `<head>` into the shadow root, "migrate" mode sacrifices some performance in favor of backwards compatibility. In short, one of the major differences between synthetic and native shadow DOM is largely eliminated, because SLDS styles (or other global stylesheets added by the page author, e.g. [static resources](https://help.salesforce.com/s/articleView?language=en_US&id=sf.os_apply_custom_styling_to_omniscripts_with_static_resources_19016.htm&type=5)) can apply to elements within the shadow DOM.
+
+This "style copying" technique is based on the [open-stylable polyfill](https://github.com/nolanlawson/open-stylable), which itself is based on the ["open-stylable" proposal](https://github.com/WICG/webcomponents/issues/909).
+
+As a performance optimization, the `<style>`s/`<link>`s will be read _once_ from the `<head>` upon encountering the first
+component in "migrate" mode. Also, global styles added by the LWC engine itself (e.g. from synthetic or top-level light-DOM components) will be skipped.

```

Comment:
This is somewhat risky, since component authors may be relying on this (e.g. a light-DOM top-level component styling synthetic shadow components). We can also consider relaxing this in a v2.

Reply:
I removed this technical detail; we have to remain flexible on how we do this.